### DISCUSSION TITLE
Can I "Resume Training" with new training labels added to the dataset?

### DISCUSSION THREAD

**Discussion by amblypatty** [talmolab/sleap#1597] [Help!] [answered]
Hi SLEAP,

I would like some clarification about training and retraining a model. Say, I have annotated 100 frames and trained a model using the 100 training frames and generated 50 predicted suggestions on unlabeled frames from the dataset. Then, I fix the predictions which adds them to the user annotated frames for training. So, now I have 150 annotated frames. 

When I make a training configuration and load a training_config.json file from the previously trained model (above, with the 100 frames), there is an option to "Resume Training" near the top left corner of the training configuration window. I see that if the Resume Training box is checked, the model configuration details become non-editable (the augmentation hyperparameters are still editable) and the [model].json file contains a definition for the `base_checkpoint` that is the filepath to the trained model folder. These are the only change I am noticing that is different in the [model].json file, please correct me if I am missing something. 

If I have the Resume Training file checked in the training configuration with 150 labeled frames (100 previous, 50 new), then does the second execution of training from this updated training package include the new frames? 

When running the new training package, it seems like the new frames are included in the training and validation sets (when validation split is 0.1):

`INFO:sleap.nn.training:  Splits: Training = 135 / Validation = 15`

This is great, but if that is the case, then I am not sure what the difference is between when the Resume Training box is checked and when it isn't checked when selecting a training_config file. 

Can I get some clarification about this please? What is the `base_checkpoint` doing, and does it affect how I should adjust any other hyperparameters, like the learning rate, for the consecutive executions of training the model?

Best,
Patrick

**Comment by talmo** [comment, marked as answer]
Hi @amblypatty,

Thanks for the question! This is a relatively new feature and I'm totally beginning to appreciate that we should probably add more information on it to our documentation.

I'm going to plagiarize myself from #1601:

> When you have other models in the `models/` subfolder in the same directory as your SLP file, they'll appear in the dropdown in the **Run Training** window, with a "**[Trained]**" prefix next to each model’s training_config.json.
> 
> You can use this feature to quickly reuse the settings you had configured previously, but you'll be training a completely new model.
> 
> When you select "**Use Trained Model**", all the fields will be grayed out no training will occur for that model and SLEAP will just use the previously trained model.
> 
> If you select "**Resume Training**", however, what SLEAP does is to train a new model, but it initializes it with the previous model weights. This is what is known as "transfer learning", or essentially reusing part of what it learned in the previous run as a starting point. This can help when you have a model trained on a different but related dataset that can bootstrap your model optimization. It does _not_, however, have any guarantee that it'll "remember" the old data nor work very well on that data. To train a model that works well on both datasets, you have to include both datasets in the same SLEAP project.


In summary, the **Resume Training** option will initialize the model with weights from the previous training run, which just changes the starting conditions of the model, but still trains on the entire labels dataset in its current state.

**If you're doing multiple rounds of training on the same dataset**, this might be useful because it might result in faster convergence since the model starts off in a state where it already works reasonably well on the previous subset of the data. On the other hand, it might lead to worse results than if you had trained from scratch (without checking **Resume Training**) if it wasn't performing well previously or if it had landed on a bad local minimum that just happened to work well on the previous subset.

**If you're starting a new project, but you have a trained model on a different but similar dataset**, this might be useful because it can leverage information learned from another (potentially larger) dataset that resulted in a model that already knows what your animals and their body parts roughly look like, even if the imaging conditions are different. Another common usecase where this can help is when you've trained another model on very similar looking data but with a different skeleton and you want to reuse the knowledge the model has about the other skeleton configuration, but update it to the new one in your current project.


Let me know if that makes sense or if you have any questions!!

Cheers,

Talmo

**Reply by amblypatty** [reply]
Hi @talmo,

Thank you for this answer. It is quite clear now. I suppose I should have been retraining the model multiple times rather than training a new model with the same configuration after adding more training labels. 

One thing I have noticed though with retraining on my project, is that I have needed to decrease the `initial_learning_rate` when I select **Resume Training** or else the model takes many more epochs before showing signs of descending. I suppose another option would be to decrease the `plateau_patience` and/or the `plateau_min_delta` for the **learning rate schedule**. My guess is that it would stop training too early by decreasing the learning rate too fast to the `min_learning_rate`. I have not tested adjusting the learning rate schedule, but am I thinking about this correctly?

Best,
Patrick
